### ISSUE TITLE
A dict word once with a subscript and once without

### ISSUE THREAD

**Issue by ampli** [opencog/link-grammar#667]
From http://www.link.cs.cmu.edu/link/dict/introduction.html:
> If a word is listed more than once with the same subscript, or if it listed once with a subscript and once without, the parser will generate a warning message and will ignore one of the entries.

Also, in a comment in `tokenize.c`:
>         The dictionary reader does not allow you to have two words that
>         match according to the criterion below.  (so you can't have
>        "dog.n" and "dog")

Currently this is enforced (in `insert_list()`) **only** if the subscripted word appear before the unsubscripted one. This is because `abridged_lookup_list()` doesn't fetch the unsubscripted word is a subscripted one is given.

There are now several words in the dict that doesn't obey this rule.
For example, `a` appears before `a.eq`, and thus is not detected. (You commented there: `% XXX why doesn't this clash with a.eq ??`)

Other words with this kind of clash: . me lying if thy seriously

**Question**: What is the reason of this rule?
If it is needed, then the check (and the dict) should be fixed.
If it is unneeded, then the check should be removed, and the dict read will then become much faster (30% for **en**).

**Comment by linas** [comment]
I cannot think of a good reason for this rule. I can think of a mediocre reason: it prevents the dictionary maintainer from accidentally creating multiple entries for the same word, and the resulting confusion from that situation.

For example, having two entries for "dog" might be confusing. By contrast having "dog.n" and "dog.v" makes it clear that one entry is meant to be a noun, the other a verb.  This makes it a bit more fool-proof for the dictionary maintainer.

I dunno. I kind-of don't care. Currently, I'm the only fool that needs to be fool-proofed - no one else is authoring these dicts.  It might be nice to have this check being optional, but its hard for me to imagine that I'd use this check very much.

**Comment by linas** [comment]
I fixed some but not all of the trouble spots in #668 I couldn't get motivated to fix  ". me seriously"

**Comment by linas** [comment]
I also removed the text "If a word is listed more than once with the same subscript, or if it listed once with a subscript and once without, the parser will generate a warning message and will ignore one of the entries."

**Comment by ampli** [comment]
> I also removed the text "If a word is listed more than once with the same subscript, or if it listed once with a subscript and once without, the parser will generate a warning message and will ignore one of the entries."

So, should we keep checking this rule in every dict read?

Note also that a similar comment exists in [tokenize.c:1089](https://github.com/opencog/link-grammar/blob/50532b7bfd34845efbf48defa22da09b0dcf3e39/link-grammar/tokenize/tokenize.c#L1089), with references to it later in this comment. This comments also refers to another obsolete stuff  (in more than one place) - the ability to put a subscripted word in a sentence (not supported now).

I propose the following:
- Fix `insert_list()` to check for this rule correctly, i.e. independent of which word come first.
- But, in order to save dict read overhead, perform this check only in verbosity=103 (verbosity>=100 is enabled only at the exact specific verbosity).
- In addition, add (to the `debug` directory) a script to perform a dict "static check", that will initially invoke `link-parser -v=103`, but eventually more checks can be added to this script (connectors which cannot be matched, regex labels not in the dict , that if `4.0.dict.m4` exists than `4.0.dict` is up to date, unused macro definitions) with an easy way for warning suppression. If all the unneeded warnings are suppressed (e.g. that <dictionary-version-number> is unused) than this script can be invoked as part of the tests.

**Comment by linas** [comment]
Yes, that proposal sounds good to me.

**Comment by ampli** [comment]
Maybe the reason of this rule was to allow removing arbitrary single words from the dict (the unused `delete_dictionary_words()`).

Anyway, I investigated the code some more and found that if I remove the said time-consuming check in `insert_list()`,  then if a word exists this generates an error in `insert_dict()` (an error that could not happen before), and a fatal dictionary read error with a memory leak happens soon.

The reason of this fatal dictionary error is that `link_advance()` is used for constructing the error message. However, if this problem is fixed, a crash happens in dsw_compression() if a word is not actually inserted (because a NULL Dict_node is inserted instead).

At this point it seems complex to undo the insertion. But I can avoid the fatal error if the duplicate word is inserted anyway (of course still with an appropriate error message). But we get a duplicate word in the directory (which still functions normally - with both sets of disjuncts for the word).

If desired,  with some more code this can be turned into a fatal dictionary error.

Alternatively, I can just insert the duplicate word with a null expression and generate an error message saying it is ignored, like now, (and still avoid the costly check before the insertion try). I think this is the best option.

Please advise.
(The rest of the solution is as I proposed before - optional full check in the code on a certain verbosity level + external dict analyzer program.)

**Comment by ampli** [comment]
I said in the post above:
> Alternatively, I can just insert the duplicate word with a null expression and generate an error message saying it is ignored, like now, (and still avoid the costly check before the insertion try). I think this is the best option.

I implemented it and it seems fine. I will send a PR soon (after finishing some more dict read speedups).

**Comment by linas** [comment]
> Please advise.
The proposal seems fine.

**Comment by ampli** [comment]
If you currently invoke `link-parser`, it reports on some dictionary problems that happened in your last fix:
```
link-grammar: Info: Dictionary found at ../en/4.0.dict
link-grammar: Warning: The word "cross-fertilized.v-d" found near line 3766 of en/4.0.dict matches the following words:	cross-fertilized.v-d
	This word will be ignored.
link-grammar: Warning: The word "cross-fertilizing.v" found near line 3781 of en/4.0.dict matches the following words:	cross-fertilizing.v
	This word will be ignored.
link-grammar: Warning: The word "cross-fertilizing.g" found near line 3784 of en/4.0.dict matches the following words:	cross-fertilizing.g
	This word will be ignored.
```

Each words needs has be removed from one of the files in which it appears, but I don't know which of them:
```
data/en/words/words.v.2.3:	cross-fertilized.v-d
data/en/words/words.v.4.3:	cross-fertilized.v-d

data/en/words/words.v.2.5:	cross-fertilizing.g
data/en/words/words.v.4.5:	cross-fertilizing.g

data/en/words/words.v.2.4:	cross-fertilizing.v
data/en/words/words.v.4.4:	cross-fertilizing.v
```


This also may indicate a problem:
```
data/en/words/words.v.4.2:	cross-fertilizes.v
data/en/words/words.v.2.2:	cross-fertilizes
```

**Comment by ampli** [comment]
I finished the patch for the optional base-word/unsubscripted-base-word check.
I made it happen in verbosity > 1 (because it indicates a "debug" invocation in which we don't care about the dict read overhead).

I also added a dictionary directive to suppress dictionary warnings:
```
/* A dictionary directive to suppress dictionary check warnings.
 * For example:
 * %suppress: DUP-BARE SOME-OTHER ... optional comment
 * The chosen symbols for warning suppression should not overlap,
 * e.g. if there is a symbol DUP-BARE, there should no symbol "DUP".
 * Their effect is until the ending ';' of the following expression.
 */
#define SUPPRESS "suppress: "
#define DUP_BASE "DUP-BASE" /* Allow a base-word + subscripted-same-word. */
```

Usage example:
```
%suppress: DUP-BASE (for a.eq)
A.eq B.eq C.eq D.eq E.eq F.eq G.eq H.eq I.eq J.eq K.eq L.eq M.eq
...
  EQt+ or EQt-;
```

There are currently 2 drawbacks (at least...) with this definition:
1. It is for everything in the definition that follows.
    If desired, something like  `DUP-BASE:a.eq` can be implemented.
2. If a warning suppression (not this one) is needed for a group of definitions, it needs to be repeated.
    If ever needed, a `%suppress: { ...` and `%suppress: }` can easily be implemented.

In verbosity=103 (an exact verbosity) I disregarded the `DUP-BASE` suppression in order to provide a full report.

P.S. A duplicate specific word is considered a problem in any case and is always ignored (it is not possible to suppress that).

Should I remove now the corresponding comment in [tokenize.c:1089](https://github.com/opencog/link-grammar/blob/50532b7bfd34845efbf48defa22da09b0dcf3e39/link-grammar/tokenize/tokenize.c#L1089)?

**Comment by linas** [comment]
> %suppress: DUP-BASE 

overkill. I don't think I'd ever use this -- its just easier to alter the entry of `a` and turn it into `a.p` if it became urgentt.

> remove comment in tokenize.c:1089

Yes.

**Comment by ampli** [comment]
>%suppress: DUP-BASE
>>overkill. I don't think I'd ever use this -- its just easier to alter the entry of a and turn it into a.p if it became urgentt.

I simply didn't want to get warnings for the current known such clashes - after I fixed the bug in the current check, that allows e.g. `a` before `a.eq` but not vise versa.
Else all of this clash check is unuseful because there is no easy way to find new such clashes as they will be intermixed with the known ones in the clash report output. Also, with no suppression mechanism, a static dictionary check that fails on dict problems cannot check this at all.

I already finished implementing it, and we can consider these possibilities:

1.  Use this patch as is, because it implements a check that could be considered useful, and also implements a general warning suppression facility that can be useful for a dict static check.
2. Altogether discard the base-word/subscripted-base-word "clash" check, because the dict currently contains several such clashes, and the current check (before my fix) is incorrect - it depends on the order of the words. (Of course in that case still leave in the duplicate strict-word check/ignore.)
